File: backend/app/schemas/group.py
"""Group and group-member Pydantic schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Group CRUD
# ---------------------------------------------------------------------------


class GroupCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=500)
    avatar_url: str | None = Field(default=None, max_length=512)


class GroupUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=500)
    avatar_url: str | None = Field(default=None, max_length=512)


class GroupRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    description: str | None = None
    owner_id: UUID
    avatar_url: str | None = None
    invite_code: str
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# Members
# ---------------------------------------------------------------------------


class GroupMemberRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    user_id: UUID
    username: str
    display_name: str | None = None
    avatar_url: str | None = None
    role: str          # 'admin' | 'member'
    joined_at: datetime


class GroupWithMembers(GroupRead):
    members: list[GroupMemberRead] = []
    chat_id: UUID | None = None   # the associated group chat


# ---------------------------------------------------------------------------
# Invite
# ---------------------------------------------------------------------------


class JoinByCodeRequest(BaseModel):
    invite_code: str = Field(min_length=1, max_length=16)


class RegenerateInviteResponse(BaseModel):
    invite_code: str

File: mobile/lib/features/groups/domain/entities/group.dart
import 'package:equatable/equatable.dart';

class GroupMember extends Equatable {
  const GroupMember({
    required this.userId,
    required this.username,
    this.displayName,
    this.avatarUrl,
    required this.role,
    required this.joinedAt,
  });

  final String userId;
  final String username;
  final String? displayName;
  final String? avatarUrl;
  final String role; // 'admin' | 'member'
  final DateTime joinedAt;

  bool get isAdmin => role == 'admin';
  String get displayLabel => displayName ?? username;

  @override
  List<Object?> get props =>
      [userId, username, displayName, avatarUrl, role, joinedAt];
}

class Group extends Equatable {
  const Group({
    required this.id,
    required this.name,
    this.description,
    required this.ownerId,
    this.avatarUrl,
    required this.inviteCode,
    required this.createdAt,
    required this.updatedAt,
    this.members = const [],
    this.chatId,
  });

  final String id;
  final String name;
  final String? description;
  final String ownerId;
  final String? avatarUrl;
  final String inviteCode;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<GroupMember> members;
  final String? chatId;

  int get memberCount => members.length;
  bool isOwner(String userId) => ownerId == userId;
  bool isAdmin(String userId) =>
      members.any((m) => m.userId == userId && m.isAdmin);

  @override
  List<Object?> get props => [
        id, name, description, ownerId, avatarUrl,
        inviteCode, createdAt, updatedAt, members, chatId,
      ];
}
